Transient cylinder example (#500)
lethe-cfd/lethe: doc/source/examples/incompressible-flow/2d-transient-flow-around-cylinder/2d-transient-flow-around-cylinder.rst

======================================
2D Transient flow around a cylinder
======================================

This example corresponds to a transient flow around a fixed cylinder at a high Reynolds number.

Features
---------

> - Solver: ``gls_navier_stokes_2d`` (with Q2-Q1)

Reviewer: ```suggestion
- Solver: ``gls_navier_stokes_2d`` (with Q2-Q1 elements)
```

.. code-block:: text

    subsection simulation control
        set method                = bdf2
        set output name           = cylinder-output
        set output frequency      = 1
        set output path           = ./Re200/
        set time end              = 200.0
        set time step		          = 0.05
        set subdivision           = 1
    end

FEM interpolation
~~~~~~~~~~~~~~~~~

The interpolation orders for the velocity and pressure are set to Q2-Q1 in the ``FEM`` subsection:

.. code-block:: text

    subsection FEM
        set velocity order  = 2
        set pressure order  = 1
    end

Mesh
~~~~~

The initial mesh is generated with `Gmsh <https://gmsh.info/#Download>`_ and imported as described in  `2D Flow around a cylinder <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-flow-around-cylinder/2d-flow-around-cylinder.html>`_.

.. code-block:: text

    subsection mesh
        set type                = gmsh
        set file name           = cylinder_structured.msh
        set initial refinement  = 1
    end

Mesh adaptation control
~~~~~~~~~~~~~~~~~~~~~~~

> While the discretization in the wake of the cylinder has less impact on the forces acting on the cylinder wall than the boundary layer discretization, it is interesting to well resolve the wake to capture the von Kármán vortex street pattern. Therefore, to adapt the mesh in the boundary layer and the wake as the vortices are shed, a non-uniform mesh adaptation is performed at each time step and the parameters are specified using the ``Mesh Adaptation Control`` subsection:

Reviewer: ```suggestion
While the discretization in the wake of the cylinder has less impact on the forces acting on the cylinder wall than the boundary layer discretization, it is interesting to resolve the wake to capture the von Kármán vortex street pattern. Therefore, to adapt the mesh in the boundary layer and the wake as the vortices are shed, a non-uniform mesh adaptation is performed at each time step and the parameters are specified using the ``Mesh Adaptation Control`` subsection:
```

.. code-block:: text

   subsection mesh adaptation
       set type                   = kelly

>        set variable               = pressure

Reviewer: It would be better if you could add a sentence to explain why you choose pressure as the refinement parameter.

Author: I added : 

`we have observed that the refinement has less tendency to follow the vortices as they move through the wake with the pressure as the refinement indicator than if we select the velocity.`

 is this better?

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The `Initial Condition <https://lethe-cfd.github.io/lethe/parameters/cfd/initial_conditions.html>`_ and `Boundary Conditions <https://lethe-cfd.github.io/lethe/parameters/cfd/boundary_conditions_cfd.html>`_ are defined as in `2D Flow around a cylinder <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-flow-around-cylinder/2d-flow-around-cylinder.html>`_.

.. code-block:: text

    subsection initial conditions
        set type = nodal
        subsection uvwp
                set Function expression = 1; 0; 0
        end
    end

    subsection boundary conditions
        set number                  = 3
        subsection bc 0
            set type              = noslip
        end
        subsection bc 1
            set type              = function
            subsection u
                set Function expression = 1
            end
            subsection v
                set Function expression = 0
            end
            subsection w
                set Function expression = 0
            end
        end
        subsection bc 2
            set type              = slip
        end
    end

Physical Properties
~~~~~~~~~~~~~~~~~~~

> The Reynolds number must be high enough to capture a transient flow and study the evolution of the drag and lift coefficients in time. Therefore, we set Re = 200 through the value of the kinematic viscosity in the same manner as for the `2D Lid-driven cavity flow <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-lid%E2%80%90driven-cavity-flow/lid%E2%80%90driven-cavity-flow.html>`_. Since :math:`U_\infty = 1` and the :math:`D = 1`, we have :math:`Re=\frac{1}{\nu}`, where :math:`\nu` is the kinematic viscosity.

Reviewer: ```suggestion
The Reynolds number must be high enough to create a transient flow and evolution of the drag and lift coefficients in time. Therefore, we set Re = 200 through choosing a kinematic viscosity in the same manner as for the `2D Lid-driven cavity flow <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-lid%E2%80%90driven-cavity-flow/lid%E2%80%90driven-cavity-flow.html>`_. Since :math:`U_\infty = 1` and the :math:`D = 1`, we have :math:`Re=\frac{1}{\nu}`, where :math:`\nu` is the kinematic viscosity.
```

        subsection fluid 0
          set kinematic viscosity            = 0.005
        end
    end

Forces
~~~~~~

Since we want to study the time evolution of the drag and lift coefficients, the force acting on the boundaries must be computed. We thus use the ``forces`` subsection:

.. code-block:: text

   subsection forces
      set verbosity             = verbose
      set calculate force       = true
      set calculate torque      = false
      set force name            = force
      set output precision      = 10
      set calculation frequency = 1
      set output frequency      = 1
   end

As we set ``calculation frequency`` to 1, the forces on each boundary are computed at each time step and written in the file specified by the field ``force name``.

.. note::

  The drag and lift coefficients are obtained with the forces acting on the wall of the cylinder (i.e., ``f_x`` and ``f_y``  written in the file ``forces.00.dat``) :

  .. math::

    C_D = \frac{2 f_x}{\rho U_\infty^2 D} \text{ and } C_L = \frac{2 f_y}{\rho U_\infty^2 D}

  where :math:`\rho = 1`. This way, we can obtain the evolution in time of both coefficients.

.. warning::

>   The computational cost of writing this output file at each time step by setting ``output frequency`` to 1 can be significant, as explained in `Force and torque calculation <https://lethe-cfd.github.io/lethe/parameters/cfd/force_and_torque.html>`_. However, in the present example, the computational cost is still fairly low so ``output frequency = 1`` is used.

Reviewer: ```suggestion
  The computational cost of writing this output file at each time step by setting ``output frequency`` to 1 can be significant, as explained in `Force and torque calculation <https://lethe-cfd.github.io/lethe/parameters/cfd/force_and_torque.html>`_. However, in the present example, the computational cost is still fairly low; so ``output frequency = 1`` is used.
```

Running the simulation
----------------------
The simulation is launched in parallel using 10 CPUs, as explained in `2D Transient flow around an Ahmed body <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-transient-around-ahmed-body/2d-transient-around-ahmed-body.html>`_ :

.. code-block:: text

  mpirun -np 10 gls_navier_stokes_2d cylinder.prm

.. warning::

>   The estimated time to simulate 200 seconds is about 2 hours 5 minutes with 10 CPUs.

Reviewer: ```suggestion
  The estimated time to simulate 200 seconds is about 2 hours and 5 minutes on 10 CPUs.
```

Results
-------

> The time evolution of the drag and lift coefficients is obtained from a Gnuplot script available in the example folder:

Reviewer: Can you mention how to call this script and the post-processing script?

Author: It's added :)

.. image:: images/CL-CD.png
    :alt: CD and CL evolution in time
    :align: center
    :name: CD-CL

> Using the FFT of the CL for the last 100 seconds, we can obtain the frequency :math:`f_v` at which the vortices are shed :

Reviewer: Does FFT stand for fast-Fourier transform? Mention it.

Author: It's added!

.. math::
 S_t = \frac{D f_v}{U_\infty}

where :math:`D` is the diameter of the cylinder, :math:`f_v` is the frequency of the shedding and :math:`U_\infty` is the upstream velocity.

Parameter file
--------------

Simulation control
~~~~~~~~~~~~~~~~~~
This example uses a 2nd order backward differentiation (``method = bdf2``) for the time integration scheme. The simulation time is set to 200 seconds with the ``time end`` parameter and a time step of 0.05 seconds is used (``time step = 0.05``).

.. code-block:: text

    subsection simulation control
        set method                = bdf2
        set output name           = cylinder-output
        set output frequency      = 1

>         set output path           = ./Re200/

Author: I create the folder (with a dummy file in it) in the example directory since there are many output files. I followed the structures of other examples to do so :)

  where :math:`\rho = 1`. This way, we can obtain the evolution in time of both coefficients.

.. warning::

  The computational cost of writing this output file at each time step by setting ``output frequency`` to 1 can be significant, as explained in `Force and torque calculation <https://lethe-cfd.github.io/lethe/parameters/cfd/force_and_torque.html>`_. However, in the present example, the computational cost is still fairly low so ``output frequency = 1`` is used.

Running the simulation
----------------------
The simulation is launched in parallel using 10 CPUs, as explained in `2D Transient flow around an Ahmed body <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-transient-around-ahmed-body/2d-transient-around-ahmed-body.html>`_ :

.. code-block:: text

  mpirun -np 10 gls_navier_stokes_2d cylinder.prm

.. warning::

  The estimated time to simulate 200 seconds is about 2 hours 5 minutes with 10 CPUs.

Results
-------

The time evolution of the drag and lift coefficients is obtained from a Gnuplot script available in the example folder by launching:

.. code-block:: text

>   gnuplot postprocess.gnu

Author: I added precision on how to run both scripts :)

2D Transient flow around a cylinder
======================================

This example corresponds to a transient flow around a fixed cylinder at a high Reynolds number.

Features
---------

- Solver: ``gls_navier_stokes_2d`` (with Q2-Q1)
- Transient problem
- Usage of Gnuplot and Python scripts for the data post-processing

Location of the example
------------------------

- Parameter file: ``/examples/incompressible_flow/2d_transient-flow_around_cylinder/cylinder.prm``
- Mesh file: ``/examples/incompressible_flow/2d_transient-flow_around_cylinder/cylinder_structured.msh``
- Geometry file: ``/examples/incompressible_flow/2d_transient-flow_around_cylinder/cylinder_structured.geo``

Description of the case
-----------------------
We simulate the flow around a fixed cylinder with a constant upstream fluid velocity. We re-use the geometry and mesh presented in `2D Flow around a cylinder <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-flow-around-cylinder/2d-flow-around-cylinder.html>`_, which were taken from Blais et al. [1]:

.. image:: images/geometry_description.png
    :alt: The geometry
    :align: center
    :name: geometry_description

> The flow field features a stable laminar boundary layer at the cylinder leading edge and a recirculation zone behind it formed by two unstable vortices of opposite signs. These vortices successively detach from the cylinder in a periodic manner (vortex shedding), leading to the generation of the von Kármán vortex street pattern in the wake. This vortex shedding causes a fluctuating pressure force acting on the cylinder, resulting in oscillations of the drag and lift coefficients in time. The frequency of vortex shedding is related to the Strouhal number:

Reviewer 2: Very nice, loved this

.. math::
 S_t = \frac{D f_v}{U_\infty}

where :math:`D` is the diameter of the cylinder, :math:`f_v` is the frequency of the shedding and :math:`U_\infty` is the upstream velocity.

Parameter file
--------------

Simulation control
~~~~~~~~~~~~~~~~~~
This example uses a 2nd order backward differentiation (``method = bdf2``) for the time integration scheme. The simulation time is set to 200 seconds with the ``time end`` parameter and a time step of 0.05 seconds is used (``time step = 0.05``).

.. code-block:: text

    subsection simulation control
        set method                = bdf2
        set output name           = cylinder-output
        set output frequency      = 1
        set output path           = ./Re200/
        set time end              = 200.0
        set time step		          = 0.05

>         set subdivision           = 1

Reviewer 2: ```suggestion
        set subdivision           = 1
```
Do you used subdivisions for the high-order rendering of Q2-Q1 results? if so explain it a bit, if not, you can remove this line.

Author: I removed it !

    end

FEM interpolation
~~~~~~~~~~~~~~~~~

The interpolation orders for the velocity and pressure are set to Q2-Q1 in the ``FEM`` subsection:

.. code-block:: text

    subsection FEM
        set velocity order  = 2
        set pressure order  = 1
    end

Mesh
~~~~~

The initial mesh is generated with `Gmsh <https://gmsh.info/#Download>`_ and imported as described in  `2D Flow around a cylinder <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-flow-around-cylinder/2d-flow-around-cylinder.html>`_.

.. code-block:: text

    subsection mesh
        set type                = gmsh
        set file name           = cylinder_structured.msh
        set initial refinement  = 1
    end

Mesh adaptation control
~~~~~~~~~~~~~~~~~~~~~~~

> While the discretization in the wake of the cylinder has less impact on the forces acting on the cylinder wall than the boundary layer discretization, it is interesting to well resolve the wake to capture the von Kármán vortex street pattern. Therefore, to adapt the mesh in the boundary layer and the wake as the vortices are shed, a non-uniform mesh adaptation is performed at each time step and the parameters are specified using the ``Mesh Adaptation Control`` subsection:

Reviewer 2: ```suggestion
While the discretization in the wake of the cylinder has less impact on the forces acting on the cylinder wall than the boundary layer discretization, it is interesting to accurately resolve the wake to capture the von Kármán vortex street pattern. Therefore, to adapt the mesh in the boundary layer and the wake as the vortices are shed, a non-uniform mesh adaptation is performed at each time step and the parameters are specified using the ``Mesh Adaptation Control`` subsection:
```

        subsection fluid 0
          set kinematic viscosity            = 0.005
        end
    end

Forces
~~~~~~

Since we want to study the time evolution of the drag and lift coefficients, the force acting on the boundaries must be computed. We thus use the ``forces`` subsection:

.. code-block:: text

   subsection forces
      set verbosity             = verbose
      set calculate force       = true
      set calculate torque      = false
      set force name            = force
      set output precision      = 10
      set calculation frequency = 1
      set output frequency      = 1
   end

As we set ``calculation frequency`` to 1, the forces on each boundary are computed at each time step and written in the file specified by the field ``force name``.

.. note::

  The drag and lift coefficients are obtained with the forces acting on the wall of the cylinder (i.e., ``f_x`` and ``f_y``  written in the file ``forces.00.dat``) :

  .. math::

    C_D = \frac{2 f_x}{\rho U_\infty^2 D} \text{ and } C_L = \frac{2 f_y}{\rho U_\infty^2 D}

  where :math:`\rho = 1`. This way, we can obtain the evolution in time of both coefficients.

.. warning::

>   The computational cost of writing this output file at each time step by setting ``output frequency`` to 1 can be significant, as explained in `Force and torque calculation <https://lethe-cfd.github.io/lethe/parameters/cfd/force_and_torque.html>`_. However, in the present example, the computational cost is still fairly low so ``output frequency = 1`` is used.

Reviewer 2: I don't think it's a good practice to write it every iteration. I would use this as an opportunity to write it every 10 or something, but i'm open minded here. :)

The obtained values of the drag and lift coefficients as well as the Strouhal number are compared to some results of the literature :

.. list-table::
   :widths: 20 20 20 20
   :header-rows: 1

   * - Study
     - :math:`C_D`
     - :math:`C_L`
     - :math:`S_t`
   * - Lethe example
     - 1.392 :math:`\pm` 0.048
     - -0.006 :math:`\pm` 0.072
     - 0.2
   * - Lethe Sharp [2]
     - 1.395 :math:`\pm` 0.047
     - :math:`\pm` 0.071
     - 0.2
   * - Braza et al. [3]
     - 1.400 :math:`\pm` 0.050
     - :math:`\pm` 0.075
     - 0.2

Using Paraview the following velocity and pressure fields can be visualized in time:

.. image:: images/cylinderVelocity.gif
    :alt: Velocity profile
    :align: center
    :name: velocity

.. image:: images/cylinderPressure.gif
    :alt: Pressure profile
    :align: center
    :name: pressure

Possibilities for extension
----------------------------
- Study the vortex shedding of other bluff bodies.
- Increase the Reynolds number to study a completely turbulent wake and the drag crisis phenomenon.

> - Repeat the same example in 3D for a cylinder/sphere and study the effect on the drag and lift forces.

Reviewer 2: I would add something like:
Investigate the impact of the time-step and the time-stepping scheme (e..g sdirk 3 or bdf 3)
